src/cursor/session.rs

```diff
@@ -347,11 +347,13 @@ impl<T> SessionCursor<T> {
 
 impl<T> SessionCursor<T> {
     fn mark_exhausted(&mut self) {
-        self.state.as_mut().unwrap().exhausted = true;
+        if let Some(state) = self.state.as_mut() {
+            state.exhausted = true;
+        }
```

Comment:
I think we can just remove the call to `mark_exhausted` in `with_type` (since it'll be a no-op with this change) and keep this method as-is.